### PR TITLE
improve conv2d

### PR DESCRIPTION
**Pull request by lm-konda**
improve loop optimization.

### CODE REVIEW

File: dlk/python/dlk/templates/src/func/conv2d.cpp [hunk 1]
```diff
@@ -102,7 +99,7 @@ void conv1x1_kn2row(T input[],
 
    assert(p.input_height > 0);
    assert(p.input_width > 0);
-
```
Comment (tkng) [inline dlk/python/dlk/templates/src/func/conv2d.cpp]: Please remove this kind of white space changes.

File: dlk/python/dlk/templates/src/func/conv2d.cpp [hunk 2]
```diff
@@ -134,24 +131,22 @@ void conv_general(
       for(T_UINT ki = 0; ki < p.kernel_height; ki++)
       {
         T_INT row = (wi * p.stride_along_height) - p.padding + ki;
-        inside_row = (row >= 0 && row < (T_INT) p.input_height);
-
+	if (row < 0 || row >= p.input_height) { current_kernel_index += p.kernel_width * p.kernel_depth; continue; }
```
Comment (tkng) [inline dlk/python/dlk/templates/src/func/conv2d.cpp]: indentation looks wrong.

File: dlk/python/dlk/templates/src/func/conv2d.cpp [hunk 3]
```diff
@@ -134,24 +131,22 @@ void conv_general(
...
+
```
Comment (tkng) [inline dlk/python/dlk/templates/src/func/conv2d.cpp]: Please remove this kind of white space changes.

File: dlk/python/dlk/templates/src/func/conv2d.cpp [hunk 4]
```diff
@@ -172,14 +167,13 @@ void convolution(
...
-    int kernels_size = p.kernel_height * p.kernel_width * p.kernel_depth * p.output_channels;
     conv1x1_kn2row(input, kernels, output, p);
     return;
   } else if (p.kernel_height == 3 && p.kernel_width == 3 && p.padding == 1) {
     int kernels_size = p.kernel_height * p.kernel_width * p.kernel_depth * p.output_channels;
-    T* kernels_hwoi = new T[kernels_size]();
+    T* kernels_hwoi = new T[kernels_size];
     ohwi_to_hwoi(kernels, kernels_hwoi, p);
-    conv3x3_kn2row(input, kernels_hwoi, output, p);
+    conv3x3_kn2row(input, kernels_hwoi, output, p);
```
Comment (tkng) [inline dlk/python/dlk/templates/src/func/conv2d.cpp]: Please remove this kind of white space changes.

**Comment by tkng** [comment]
For this kind of performance improvement, please share the result of benchmark.

Also, plesae include the procedure to reproduce the benchmark result, so that reviewer can check the result.

File: dlk/python/dlk/templates/src/func/conv2d.cpp (hunk 1 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/src/func/conv2d.cpp]: I will fix it.

Comment (lm-konda) [reply dlk/python/dlk/templates/src/func/conv2d.cpp]: fixed

File: dlk/python/dlk/templates/src/func/conv2d.cpp (hunk 2 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/src/func/conv2d.cpp]: fixed

File: dlk/python/dlk/templates/src/func/conv2d.cpp (hunk 3 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/src/func/conv2d.cpp]: fixed

File: dlk/python/dlk/templates/src/func/conv2d.cpp (hunk 4 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/src/func/conv2d.cpp]: fixed

**Comment by lm-konda** [comment]
performance

before
```
Convolution,168176,2093,  sum:170.269ms
TotalRunTime,302676,  sum:302.676ms
```
after
```
Convolution,166709,2091,  sum:168.8ms
TotalRunTime,301169,  sum:301.169ms
```

File: dlk/python/dlk/templates/src/func/conv2d.cpp [hunk 5]
```diff
@@ -45,9 +45,6 @@ void conv3x3_kn2row(T input[],
   assert(kh == 3 && kw == 3);
   assert(MAX_SIZE_KN2ROW_BUFFER_PER_LAYER >= oc * kh * kw * ih * iw);
 
-  // need to initialize output
```
Comment (tkng) [inline dlk/python/dlk/templates/src/func/conv2d.cpp]: Is it possible to remove this `std::memset`? According to the code, it looks impossible for me.

**Comment by tkng** [comment]
Sorry, my words were not enough. Please include the following information.

- network you used
- dataset you used
- training configuration you used
- hardware you used
- how many times you run the benchmark (I want to know that whether the improvement is statistically significant.)

These're what I intended by the word "procedure to reproduce the benchmark result".

File: dlk/python/dlk/templates/src/func/conv2d.cpp (hunk 5 above)
Comment (lm-konda) [reply dlk/python/dlk/templates/src/func/conv2d.cpp]: As a result of my detailed examination, you were right. Very sorry. I fixed it.

**Comment by lm-konda** [comment]
I am almost always using the configuration below.

```
network
　LmnetV1Quantize
dataset you used
　Caltech101
training configurattion
　epoch is 2. other is default.
hardware
　DE10-nano
how many times you run the benchmark
　As much as average average value is known (around 10 or more)
```

**Comment by tkng** [comment]
> As much as average average value is known (around 10 or more)

Please tell me the average and variance, since the effectiveness of this change is not obvious for me, "speed up to 168.8ms from 170.269ms" looks that just a fluctuation.

**Comment by lm-konda** [comment]
@tkng 
You are right.
There was no significant performance improvement when remeasurement was done. 
The reason is probably because I have cut down the necessary memset.
Sorry. I withdraw this PR.

**Pull request closed without merging**